File: python/0008-string-to-integer-atoi.py
"""
LeetCode 0008: String to Integer (atoi)

Problem Statement:
Implement the myAtoi(string s) function, which converts a string to a signed 32-bit integer.

The algorithm for myAtoi(string s) is as follows:
1. Read in and ignore any leading whitespace.
2. Check if the next character is '-' or '+'. Read this character in if it is either. This determines the sign of the final result.
3. Read in next the characters until the next non-digit character or the end of the input is reached. The rest of the string is ignored.
4. Convert these digits into an integer (i.e. "123" -> 123, "0032" -> 32). If no digits were read, then the result is 0.
5. If the integer is out of the 32-bit signed integer range [-2^31, 2^31 - 1], then clamp the result so that it remains in the range.
6. Return the integer as the final result.

Example 1:
Input: s = "42"
Output: 42

Example 2:
Input: s = "   -42"
Output: -42

Example 3:
Input: s = "4193 with words"
Output: 4193

Constraints:
- 0 <= s.length <= 200
- s consists of English letters (lower-case and upper-case), digits (0-9), ' ', '+', '-', and '.'.
- -2^31 <= return value <= 2^31 - 1
"""


class StringToInteger:
    """Solution for LeetCode 0008: String to Integer (atoi)"""

    INT_MIN = -2147483648  # -2^31
    INT_MAX = 2147483647   # 2^31 - 1

    # Q1: What's the step-by-step approach to solve this problem?
    # A1: Parse string following the atoi algorithm. Time: O(n), Space: O(1)
    def my_atoi_manual(self, s: str) -> int:
        """
        Algorithm:
        1. Skip leading whitespace
        2. Check for sign (+/-)
        3. Read digits until non-digit or end
        4. Build number
        5. Clamp to 32-bit range
        """
        # Step 1: Skip leading whitespace
        i = 0
        while i < len(s) and s[i] == ' ':
            i += 1

        # If we reached the end, return 0
        if i == len(s):
            return 0

        # Step 2: Check for sign
        sign = 1
        if s[i] == '+':
            i += 1
        elif s[i] == '-':
            sign = -1
            i += 1

        # Step 3 & 4: Read digits and build number
        result = 0
        while i < len(s) and s[i].isdigit():
            digit = int(s[i])

            # Check for overflow before multiplication
            if result > self.INT_MAX // 10:
                return self.INT_MAX if sign == 1 else self.INT_MIN
            if result == self.INT_MAX // 10 and digit > 7:
                return self.INT_MAX if sign == 1 else self.INT_MIN

            result = result * 10 + digit
            i += 1

        # Apply sign
        result *= sign

        # Step 5: Clamp to 32-bit range
        if result < self.INT_MIN:
            return self.INT_MIN
        if result > self.INT_MAX:
            return self.INT_MAX

        return result

    # Q2: How can we simplify the overflow check?
    # A2: Use Python's integer flexibility, check at the end. Time: O(n), Space: O(1)
    def my_atoi_simplified(self, s: str) -> int:
        """
        Algorithm:
        1. Skip whitespace
        2. Parse sign
        3. Extract digits
        4. Clamp result
        """
        i = 0

        # Skip leading whitespace
        while i < len(s) and s[i] == ' ':
            i += 1

        # Parse sign
        sign = 1
        if i < len(s) and s[i] in ['+', '-']:
            if s[i] == '-':
                sign = -1
            i += 1

        # Extract digits
        result = 0
        while i < len(s) and s[i].isdigit():
            result = result * 10 + int(s[i])
            i += 1

        result *= sign

        # Clamp to 32-bit range
        if result < self.INT_MIN:
            return self.INT_MIN
        if result > self.INT_MAX:
            return self.INT_MAX

        return result

    # Q3: Can we use regex for pattern matching?
    # A3: Yes, but it's more complex. Time: O(n), Space: O(n) for regex pattern
    def my_atoi_regex(self, s: str) -> int:
        """
        Algorithm:
        1. Use regex to match the pattern
        2. Extract the number
        3. Clamp result
        """
        import re

        # Pattern: optional spaces, optional sign, one or more digits
        match = re.match(r'^\s*([+-]?\d+)', s)

        if not match:
            return 0

        result = int(match.group(1))

        # Clamp to 32-bit range
        if result < self.INT_MIN:
            return self.INT_MIN
        if result > self.INT_MAX:
            return self.INT_MAX

        return result

    # Q4: What are the edge cases we need to handle?
    """
    A4: Edge cases:

    1. Empty string: "" -> 0
    2. Only whitespace: "   " -> 0
    3. No digits: "+abc" -> 0
    4. Leading zeros: "00123" -> 123
    5. Sign followed by spaces: "+ 42" -> 0 (space after sign)
    6. Words after number: "4193 with words" -> 4193
    7. Only sign: "+" -> 0, "-" -> 0
    8. Overflow positive: "9999999999" -> 2147483647
    9. Overflow negative: "-9999999999" -> -2147483648
    10. Decimal point: "3.14159" -> 3
    11. Multiple signs: "++42" -> 0 (double sign)
    12. Space in middle: "1 2 3" -> 1
    """

    # Q5: How to handle whitespace correctly?
    """
    A5: Whitespace handling:

    1. Only leading whitespace is skipped
    2. Whitespace after sign stops parsing
    3. Example: "  +0 123" -> 0
    4. Example: "  -42 abc" -> -42
    5. Use isspace() for flexibility or check ' ' for strict ASCII space
    """

    # Q6: What about the sign character?
    """
    A6: Sign character rules:

    1. Only '+' or '-' is valid sign
    2. Must come after whitespace, before digits
    3. Multiple signs: "++42" -> 0 (treated as non-digit)
    4. Other characters: "*42" -> 0
    5. Sign between digits: "4-2" -> 4
    6. Only one sign is allowed
    """

    # Q7: How to detect overflow efficiently?
    """
    A7: Overflow detection strategies:

    1. Pre-check: Before result = result * 10 + digit
       - Check if result > INT_MAX // 10
       - Check if result == INT_MAX // 10 and digit > 7

    2. Post-check: After building result, compare with bounds
       - Simpler but might allow temporary overflow (not in Python)
       - Python: No native overflow, so post-check is safe

    3. For this problem: Post-check is cleaner in Python
    """

    # Q8: What's the difference between atoi and int()?
    """
    A8: Differences between atoi and int():

    1. atoi stops at first non-digit (after sign)
    2. int() raises ValueError on invalid input
    3. atoi returns 0 on error, int() raises exception
    4. atoi clamps to 32-bit, int() supports arbitrary precision
    5. atoi is more lenient (4193 with words -> 4193)
    6. int() is stricter (int("4193 with words") raises error)

    Example:
    - atoi("4193 with words") = 4193
    - int("4193 with words") -> ValueError
    """

    # Q9: Can we use state machine approach?
    # A9: Yes, state machine for cleaner parsing. Time: O(n), Space: O(1)
    def my_atoi_state_machine(self, s: str) -> int:
        """
        State machine approach:
        States: START, SIGN, NUMBER, END
        """
        state = "START"
        sign = 1
        result = 0

        for char in s:
            if state == "START":
                if char == ' ':
                    continue
                elif char in ['+', '-']:
                    sign = 1 if char == '+' else -1
                    state = "SIGN"
                elif char.isdigit():
                    result = int(char)
                    state = "NUMBER"
                else:
                    return 0
            elif state == "SIGN":
                if char.isdigit():
                    result = int(char)
                    state = "NUMBER"
                else:
                    return 0
            elif state == "NUMBER":
                if char.isdigit():
                    result = result * 10 + int(char)
                else:
                    break

        result *= sign

        # Clamp to 32-bit range
        if result < self.INT_MIN:
            return self.INT_MIN
        if result > self.INT_MAX:
            return self.INT_MAX

        return result

    # Q10: How to compare approaches?
    """
    A10: Complexity comparison:

    1. Manual parsing (Pre-check):
       - Pros: Efficient overflow handling
       - Cons: More complex overflow logic
       - Time: O(n), Space: O(1)

    2. Simplified parsing (Post-check):
       - Pros: Cleaner, easier to understand
       - Cons: Slightly less efficient for large numbers
       - Time: O(n), Space: O(1)

    3. Regex approach:
       - Pros: Concise, pattern is clear
       - Cons: Extra space for regex, overhead
       - Time: O(n), Space: O(n)

    4. State machine:
       - Pros: Very clear state transitions
       - Cons: More code, slightly more overhead
       - Time: O(n), Space: O(1)

    Best for interview: Simplified parsing - clear and concise
    """

    # Q11: What are common mistakes?
    """
    A11: Common mistakes:

    1. Not skipping leading whitespace
    2. Allowing multiple signs: "++42" should be 0
    3. Allowing space after sign: "+ 42" should be 0
    4. Not stopping at first non-digit
    5. Not handling overflow: 9999999999 should clamp
    6. Using int() directly which raises exceptions
    7. Forgetting that "3.14" should be 3
    8. Not testing edge cases thoroughly

    Testing checklist:
    - Empty string
    - Whitespace only
    - Leading zeros
    - Very large numbers
    - Negative numbers
    - Words after numbers
    - Invalid characters
    """

    # Q12: Real-world applications?
    """
    A12: Applications:

    1. String parsing: Converting string input to integers
    2. Configuration files: Parsing numeric values
    3. User input validation: Extracting numbers from input
    4. Data parsing: Reading formatted data
    5. Protocol parsing: Extracting numbers from messages
    6. Log parsing: Extracting numeric values from logs
    7. Scripting languages: Implementing atoi function
    """

    def test_my_atoi(self, method_name: str = "simplified") -> bool:
        """
        Comprehensive test suite
        """
        test_cases = [
            # (input, expected_output)
            ("42", 42),
            ("   -42", -42),
            ("4193 with words", 4193),
            ("words and 987", 0),
            ("-91283472332", -2147483648),
            ("0", 0),
            ("", 0),
            ("   ", 0),
            ("+123", 123),
            ("+-42", 0),
            ("00123", 123),
            ("3.14159", 3),
            ("+0 123", 0),
            ("9999999999", 2147483647),
            ("-9999999999", -2147483648),
            ("+", 0),
            ("-", 0),
            ("+ 42", 0),
            ("  +0 123", 0),
            ("1", 1),
            ("-1", -1),
            ("2147483647", 2147483647),
            ("-2147483648", -2147483648),
        ]

        passed = 0
        failed = 0

        method = {
            "manual": self.my_atoi_manual,
            "simplified": self.my_atoi_simplified,
            "regex": self.my_atoi_regex,
            "state_machine": self.my_atoi_state_machine,
        }.get(method_name, self.my_atoi_simplified)

        print(f"Testing {method_name} approach:")
        for input_val, expected in test_cases:
            result = method(input_val)
            status = "PASS" if result == expected else "FAIL"
            if status == "PASS":
                passed += 1
                print(f"  {status}: myAtoi(\"{input_val}\") = {result}")
            else:
                failed += 1
                print(f"  {status}: myAtoi(\"{input_val}\") = {result} (expected {expected})")

        print(f"\nResults: {passed} passed, {failed} failed")
        return failed == 0


def main():
    """Main method for testing"""
    solution = StringToInteger()

    print("=" * 60)
    print("LeetCode 0008: String to Integer (atoi)")
    print("=" * 60)

    # Example 1
    s1 = "42"
    print("\nExample 1:")
    print(f"Input: s = \"{s1}\"")
    print(f"Output (Manual):        {solution.my_atoi_manual(s1)}")
    print(f"Output (Simplified):    {solution.my_atoi_simplified(s1)}")
    print(f"Output (Regex):         {solution.my_atoi_regex(s1)}")
    print(f"Output (State Machine): {solution.my_atoi_state_machine(s1)}")

    # Example 2
    s2 = "   -42"
    print("\n" + "-" * 60)
    print("Example 2:")
    print(f"Input: s = \"{s2}\"")
    print(f"Output (Manual):        {solution.my_atoi_manual(s2)}")
    print(f"Output (Simplified):    {solution.my_atoi_simplified(s2)}")
    print(f"Output (Regex):         {solution.my_atoi_regex(s2)}")
    print(f"Output (State Machine): {solution.my_atoi_state_machine(s2)}")

    # Example 3
    s3 = "4193 with words"
    print("\n" + "-" * 60)
    print("Example 3:")
    print(f"Input: s = \"{s3}\"")
    print(f"Output (Manual):        {solution.my_atoi_manual(s3)}")
    print(f"Output (Simplified):    {solution.my_atoi_simplified(s3)}")
    print(f"Output (Regex):         {solution.my_atoi_regex(s3)}")
    print(f"Output (State Machine): {solution.my_atoi_state_machine(s3)}")

    # Example 4 - Edge case
    s4 = "words and 987"
    print("\n" + "-" * 60)
    print("Example 4 (Invalid):")
    print(f"Input: s = \"{s4}\"")
    print(f"Output (Manual):        {solution.my_atoi_manual(s4)}")
    print(f"Output (Simplified):    {solution.my_atoi_simplified(s4)}")
    print(f"Output (Regex):         {solution.my_atoi_regex(s4)}")
    print(f"Output (State Machine): {solution.my_atoi_state_machine(s4)}")

    # Example 5 - Overflow
    s5 = "-91283472332"
    print("\n" + "-" * 60)
    print("Example 5 (Overflow):")
    print(f"Input: s = \"{s5}\"")
    print(f"Output (Manual):        {solution.my_atoi_manual(s5)}")
    print(f"Output (Simplified):    {solution.my_atoi_simplified(s5)}")
    print(f"Output (Regex):         {solution.my_atoi_regex(s5)}")
    print(f"Output (State Machine): {solution.my_atoi_state_machine(s5)}")

    # Comprehensive testing
    print("\n" + "=" * 60)
    print("Running Comprehensive Tests:")
    print("=" * 60 + "\n")
    solution.test_my_atoi("simplified")

    # Performance comparison
    print("\n" + "=" * 60)
    print("Performance Comparison:")
    print("=" * 60)

    import time

    test_strings = ["42", "   -42", "4193 with words", "words and 987", "-91283472332"]
    iterations = 100000

    for s in test_strings:
        start_time = time.time()
        for _ in range(iterations):
            solution.my_atoi_manual(s)
        manual_time = (time.time() - start_time) * 1000

        start_time = time.time()
        for _ in range(iterations):
            solution.my_atoi_simplified(s)
        simplified_time = (time.time() - start_time) * 1000

        start_time = time.time()
        for _ in range(iterations):
            solution.my_atoi_regex(s)
        regex_time = (time.time() - start_time) * 1000

        start_time = time.time()
        for _ in range(iterations):
            solution.my_atoi_state_machine(s)
        state_machine_time = (time.time() - start_time) * 1000

        print(f"\nInput: \"{s}\" ({iterations} iterations)")
        print(f"  Manual approach:         {manual_time:.2f} ms")
        print(f"  Simplified approach:     {simplified_time:.2f} ms")
        print(f"  Regex approach:          {regex_time:.2f} ms")
        print(f"  State Machine approach:  {state_machine_time:.2f} ms")


if __name__ == "__main__":
    main()

File: java/0008-string-to-integer-atoi.java
import java.util.Arrays;

/**
 * LeetCode 0008: String to Integer (atoi)
 *
 * Problem Statement:
 * Implement the myAtoi(String s) function, which converts a string to a signed 32-bit integer.
 *
 * The algorithm for myAtoi(String s) is as follows:
 * 1. Read in and ignore any leading whitespace.
 * 2. Check if the next character is '-' or '+'. Read this character in if it is either.
 * 3. Read in next the characters until the next non-digit character or the end of the input is reached.
 * 4. Convert these digits into an integer (i.e. "123" -> 123, "0032" -> 32).
 * 5. If the integer is out of the 32-bit signed integer range [-2^31, 2^31 - 1], clamp the result.
 * 6. Return the integer as the final result.
 *
 * Example 1:
 * Input: s = "42"
 * Output: 42
 *
 * Example 2:
 * Input: s = "   -42"
 * Output: -42
 *
 * Example 3:
 * Input: s = "4193 with words"
 * Output: 4193
 *
 * Constraints:
 * - 0 <= s.length() <= 200
 * - s consists of English letters (lower-case and upper-case), digits (0-9), ' ', '+', '-', and '.'.
 * - -2^31 <= return value <= 2^31 - 1
 */

public class StringToInteger {

    private static final int INT_MIN = -2147483648;  // -2^31
    private static final int INT_MAX = 2147483647;   // 2^31 - 1

    /**
     * Q1: What's the step-by-step approach to solve this problem?
     * A1: Parse string following the atoi algorithm. Time: O(n), Space: O(1)
     *
     * Algorithm:
     * 1. Skip leading whitespace
     * 2. Check for sign (+/-)
     * 3. Read digits until non-digit or end
     * 4. Build number with overflow check
     * 5. Return clamped result
     */
    public int myAtoiManual(String s) {
        // Step 1: Skip leading whitespace
        int i = 0;
        while (i < s.length() && s.charAt(i) == ' ') {
            i++;
        }

        // If we reached the end, return 0
        if (i == s.length()) {
            return 0;
        }

        // Step 2: Check for sign
        int sign = 1;
        if (s.charAt(i) == '+') {
            i++;
        } else if (s.charAt(i) == '-') {
            sign = -1;
            i++;
        }

        // Step 3 & 4: Read digits and build number with overflow check
        int result = 0;
        while (i < s.length() && Character.isDigit(s.charAt(i))) {
            int digit = s.charAt(i) - '0';

            // Check for overflow before multiplication
            if (result > INT_MAX / 10) {
                return sign == 1 ? INT_MAX : INT_MIN;
            }
            if (result == INT_MAX / 10 && digit > 7) {
                return sign == 1 ? INT_MAX : INT_MIN;
            }

            result = result * 10 + digit;
            i++;
        }

        return result * sign;
    }

    /**
     * Q2: How can we simplify the overflow check?
     * A2: Check after parsing, using long for intermediate result. Time: O(n), Space: O(1)
     *
     * Algorithm:
     * 1. Skip whitespace
     * 2. Parse sign
     * 3. Extract digits into long
     * 4. Clamp result to 32-bit range
     */
    public int myAtoiSimplified(String s) {
        // Skip leading whitespace
        int i = 0;
        while (i < s.length() && s.charAt(i) == ' ') {
            i++;
        }

        // Parse sign
        int sign = 1;
        if (i < s.length()) {
            if (s.charAt(i) == '+') {
                i++;
            } else if (s.charAt(i) == '-') {
                sign = -1;
                i++;
            }
        }

        // Extract digits using long to avoid overflow temporarily
        long result = 0;
        while (i < s.length() && Character.isDigit(s.charAt(i))) {
            result = result * 10 + (s.charAt(i) - '0');

            // Early termination if already out of bounds
            if (result > INT_MAX) {
                break;
            }
            i++;
        }

        result *= sign;

        // Clamp to 32-bit range
        if (result < INT_MIN) {
            return INT_MIN;
        }
        if (result > INT_MAX) {
            return INT_MAX;
        }

        return (int) result;
    }

    /**
     * Q3: Can we use a state machine approach?
     * A3: Yes, state machine for cleaner parsing. Time: O(n), Space: O(1)
     *
     * States: START, SIGN, NUMBER, END
     */
    public int myAtoiStateMachine(String s) {
        String state = "START";
        int sign = 1;
        long result = 0;

        for (char c : s.toCharArray()) {
            if (state.equals("START")) {
                if (c == ' ') {
                    continue;
                } else if (c == '+') {
                    state = "SIGN";
                } else if (c == '-') {
                    sign = -1;
                    state = "SIGN";
                } else if (Character.isDigit(c)) {
                    result = c - '0';
                    state = "NUMBER";
                } else {
                    return 0;
                }
            } else if (state.equals("SIGN")) {
                if (Character.isDigit(c)) {
                    result = c - '0';
                    state = "NUMBER";
                } else {
                    return 0;
                }
            } else if (state.equals("NUMBER")) {
                if (Character.isDigit(c)) {
                    result = result * 10 + (c - '0');
                } else {
                    break;
                }
            }
        }

        result *= sign;

        // Clamp to 32-bit range
        if (result < INT_MIN) {
            return INT_MIN;
        }
        if (result > INT_MAX) {
            return INT_MAX;
        }

        return (int) result;
    }

    /**
     * Q4: What are the edge cases we need to handle?
     *
     * A4: Edge cases:
     * 1. Empty string: "" -> 0
     * 2. Only whitespace: "   " -> 0
     * 3. No digits: "+abc" -> 0
     * 4. Leading zeros: "00123" -> 123
     * 5. Sign followed by spaces: "+ 42" -> 0 (space after sign)
     * 6. Words after number: "4193 with words" -> 4193
     * 7. Only sign: "+" -> 0, "-" -> 0
     * 8. Overflow positive: "9999999999" -> 2147483647
     * 9. Overflow negative: "-9999999999" -> -2147483648
     * 10. Decimal point: "3.14159" -> 3
     */

    /**
     * Q5: How to handle overflow detection efficiently?
     *
     * A5: Overflow detection strategies:
     * 1. Pre-check: Before result = result * 10 + digit
     *    - Check if result > INT_MAX / 10
     *    - Check if result == INT_MAX / 10 && digit > 7
     * 2. Post-check: After parsing, compare with bounds
     * 3. Use long: Parse into long, then clamp (simpler in Java)
     */

    /**
     * Q6: Difference between atoi and Integer.parseInt()?
     *
     * A6: Differences:
     * 1. atoi stops at first non-digit (after sign)
     * 2. Integer.parseInt() throws NumberFormatException on invalid input
     * 3. atoi returns 0 on error, parseInt throws exception
     * 4. atoi is more lenient ("4193 with words" -> 4193)
     * 5. parseInt is stricter (throws exception)
     */

    /**
     * Q7: How to test edge cases thoroughly?
     *
     * A7: Testing approach:
     * 1. Create comprehensive test suite with all edge cases
     * 2. Test empty strings, whitespace, signs
     * 3. Test overflow cases
     * 4. Test boundary values
     * 5. Test mixed content (digits with words)
     */

    public boolean testMyAtoi(String methodName) {
        String[][] testCases = {
            {"42", "42"},
            {"   -42", "-42"},
            {"4193 with words", "4193"},
            {"words and 987", "0"},
            {"-91283472332", String.valueOf(INT_MIN)},
            {"0", "0"},
            {"", "0"},
            {"   ", "0"},
            {"+123", "123"},
            {"+-42", "0"},
            {"00123", "123"},
            {"3.14159", "3"},
            {"+0 123", "0"},
            {"9999999999", String.valueOf(INT_MAX)},
            {"-9999999999", String.valueOf(INT_MIN)},
            {"+", "0"},
            {"-", "0"},
            {"+ 42", "0"},
            {"  +0 123", "0"},
            {"1", "1"},
            {"-1", "-1"},
            {"2147483647", "2147483647"},
            {"-2147483648", "-2147483648"},
        };

        int passed = 0;
        int failed = 0;

        System.out.println("Testing " + methodName + " approach:");
        for (String[] testCase : testCases) {
            String input = testCase[0];
            int expected = Integer.parseInt(testCase[1]);

            int result;
            switch (methodName) {
                case "manual":
                    result = myAtoiManual(input);
                    break;
                case "state_machine":
                    result = myAtoiStateMachine(input);
                    break;
                default:
                    result = myAtoiSimplified(input);
                    break;
            }

            if (result == expected) {
                passed++;
                System.out.println("  PASS: myAtoi(\"" + input + "\") = " + result);
            } else {
                failed++;
                System.out.println("  FAIL: myAtoi(\"" + input + "\") = " + result + " (expected " + expected + ")");
            }
        }

        System.out.println("\nResults: " + passed + " passed, " + failed + " failed");
        return failed == 0;
    }

    public static void main(String[] args) {
        StringToInteger solution = new StringToInteger();

        System.out.println("=".repeat(60));
        System.out.println("LeetCode 0008: String to Integer (atoi)");
        System.out.println("=".repeat(60));

        // Example 1
        String s1 = "42";
        System.out.println("\nExample 1:");
        System.out.println("Input: s = \"" + s1 + "\"");
        System.out.println("Output (Manual):        " + solution.myAtoiManual(s1));
        System.out.println("Output (Simplified):    " + solution.myAtoiSimplified(s1));
        System.out.println("Output (State Machine): " + solution.myAtoiStateMachine(s1));

        // Example 2
        String s2 = "   -42";
        System.out.println("\n" + "-".repeat(60));
        System.out.println("Example 2:");
        System.out.println("Input: s = \"" + s2 + "\"");
        System.out.println("Output (Manual):        " + solution.myAtoiManual(s2));
        System.out.println("Output (Simplified):    " + solution.myAtoiSimplified(s2));
        System.out.println("Output (State Machine): " + solution.myAtoiStateMachine(s2));

        // Example 3
        String s3 = "4193 with words";
        System.out.println("\n" + "-".repeat(60));
        System.out.println("Example 3:");
        System.out.println("Input: s = \"" + s3 + "\"");
        System.out.println("Output (Manual):        " + solution.myAtoiManual(s3));
        System.out.println("Output (Simplified):    " + solution.myAtoiSimplified(s3));
        System.out.println("Output (State Machine): " + solution.myAtoiStateMachine(s3));

        // Example 4 - Edge case
        String s4 = "words and 987";
        System.out.println("\n" + "-".repeat(60));
        System.out.println("Example 4 (Invalid):");
        System.out.println("Input: s = \"" + s4 + "\"");
        System.out.println("Output (Manual):        " + solution.myAtoiManual(s4));
        System.out.println("Output (Simplified):    " + solution.myAtoiSimplified(s4));
        System.out.println("Output (State Machine): " + solution.myAtoiStateMachine(s4));

        // Example 5 - Overflow
        String s5 = "-91283472332";
        System.out.println("\n" + "-".repeat(60));
        System.out.println("Example 5 (Overflow):");
        System.out.println("Input: s = \"" + s5 + "\"");
        System.out.println("Output (Manual):        " + solution.myAtoiManual(s5));
        System.out.println("Output (Simplified):    " + solution.myAtoiSimplified(s5));
        System.out.println("Output (State Machine): " + solution.myAtoiStateMachine(s5));

        // Comprehensive testing
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Running Comprehensive Tests:");
        System.out.println("=".repeat(60) + "\n");
        solution.testMyAtoi("simplified");

        // Performance comparison
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Performance Comparison:");
        System.out.println("=".repeat(60));

        String[] testStrings = {"42", "   -42", "4193 with words", "words and 987", "-91283472332"};
        int iterations = 100000;

        for (String s : testStrings) {
            long startTime = System.currentTimeMillis();
            for (int i = 0; i < iterations; i++) {
                solution.myAtoiManual(s);
            }
            long manualTime = System.currentTimeMillis() - startTime;

            startTime = System.currentTimeMillis();
            for (int i = 0; i < iterations; i++) {
                solution.myAtoiSimplified(s);
            }
            long simplifiedTime = System.currentTimeMillis() - startTime;

            startTime = System.currentTimeMillis();
            for (int i = 0; i < iterations; i++) {
                solution.myAtoiStateMachine(s);
            }
            long stateMachineTime = System.currentTimeMillis() - startTime;

            System.out.println("\nInput: \"" + s + "\" (" + iterations + " iterations)");
            System.out.println("  Manual approach:        " + manualTime + " ms");
            System.out.println("  Simplified approach:    " + simplifiedTime + " ms");
            System.out.println("  State Machine approach: " + stateMachineTime + " ms");
        }
    }
}
